File: zap-niver-dashboard/src/services/templateService.ts
import { supabase } from '@/lib/store/apiClient'

export interface Template {
  id: string
  user_id: string
  name: string
  content: string
  is_active: boolean
  created_at: string
}

/**
 * Serviço para gerenciar templates de mensagens
 */
export const templateService = {
  /**
   * Busca todos os templates do usuário
   */
  async getUserTemplates(userId: string): Promise<Template[]> {
    try {
      // Usar o método order corretamente com o novo apiClient
      const result = supabase
        .from('templates')
        .select('*')
        .order('name', { ascending: true })
      
      const { data, error } = await result.execute()

      if (error) {
        console.error('Erro ao buscar templates:', error)
        throw error
      }

      // Filtrar os resultados manualmente
      const filteredData = data?.filter(template => template.user_id === userId) || []

      return filteredData
    } catch (error) {
      console.error('Erro ao buscar templates:', error)
      return []
    }
  },

  /**
   * Conta o total de templates ativos do usuário
   */
  async countActiveTemplates(userId: string): Promise<number> {
    try {
      // Usar o novo apiClient para contar templates
      const result = supabase
        .from('templates')
        .select('*')
      
      const { data, error } = await result.execute()

      if (error) {
        console.error('Erro ao contar templates ativos:', error)
        throw error
      }

      // Filtrar os resultados manualmente
      const filteredData = data?.filter(template => 
        template.user_id === userId && 
        template.is_active === true
      ) || []

      return filteredData.length
    } catch (error) {
      console.error('Erro ao contar templates ativos:', error)
      return 0
    }
  },

  /**
   * Busca um template específico pelo ID
   */
  async getTemplateById(id: string): Promise<Template | null> {
    try {
      // Usar o método select corretamente com o novo apiClient
      const result = supabase
        .from('templates')
        .select('*')
      
      const { data, error } = await result.execute()

      if (error) {
        console.error(`Erro ao buscar template ${id}:`, error)
        throw error
      }

      // Encontrar o template pelo ID manualmente
      const template = data?.find(tpl => tpl.id === id) || null

      return template
    } catch (error) {
      console.error(`Erro ao buscar template ${id}:`, error)
      return null
    }
  },

  /**
   * Cria um novo template
   */
  async createTemplate(template: Omit<Template, 'id' | 'created_at'>): Promise<Template> {
    try {
      // Usar o método insert corretamente com o novo apiClient
      const { data, error } = await supabase
        .from('templates')
        .insert(template)

      if (error) {
        console.error('Erro ao criar template:', error)
        throw error
      }

      return data
    } catch (error) {
      console.error('Erro ao criar template:', error)
      throw error
    }
  },

  /**
   * Atualiza um template existente
   */
  async updateTemplate(id: string, template: Partial<Omit<Template, 'id' | 'created_at'>>): Promise<Template> {
    try {
      // Usar o método update corretamente com o novo apiClient
      const { data, error } = await supabase
        .from('templates')
        .update(template)
        .eq('id', id)

      if (error) {
        console.error(`Erro ao atualizar template ${id}:`, error)
        throw error
      }

      return data
    } catch (error) {
      console.error(`Erro ao atualizar template ${id}:`, error)
      throw error
    }
  },

  /**
   * Exclui um template
   */
  async deleteTemplate(id: string): Promise<void> {
    try {
      // Usar o método delete corretamente com o novo apiClient
      const { error } = await supabase
        .from('templates')
        .delete()
        .eq('id', id)

      if (error) {
        console.error(`Erro ao excluir template ${id}:`, error)
        throw error
      }
    } catch (error) {
      console.error(`Erro ao excluir template ${id}:`, error)
      throw error
    }
  }
}

File: backend/app/services/template_service.py
from typing import Optional, List, Dict, Any
from datetime import datetime

from app.db.database import get_db_connection
from app.models.template import TemplateCreate, TemplateUpdate, Template, TemplateList

class TemplateService:
    """Serviço para gerenciar templates de mensagem no PostgreSQL"""
    
    @staticmethod
    async def create_template(user_id: str, template_in: TemplateCreate) -> Template:
        """
        Cria um novo template
        
        Args:
            user_id: ID do usuário proprietário do template
            template_in: Dados do template a ser criado
            
        Returns:
            Template criado
        """
        async with await get_db_connection() as conn:
            data = template_in.dict()
            template_id = await conn.fetchval(
                """
                INSERT INTO templates (user_id, name, content, is_active, created_at, updated_at)
                VALUES ($1, $2, $3, TRUE, $4, $4)
                RETURNING id
                """,
                user_id,
                data["name"],
                data["content"],
                datetime.utcnow()
            )
            row = await conn.fetchrow("SELECT * FROM templates WHERE id=$1", template_id)
            return Template(**dict(row))
    
    @staticmethod
    async def get_template(user_id: str, template_id: str) -> Optional[Template]:
        """
        Obtém um template pelo ID
        
        Args:
            user_id: ID do usuário proprietário do template
            template_id: ID do template
            
        Returns:
            Template ou None se não encontrado
        """
        async with await get_db_connection() as conn:
            row = await conn.fetchrow(
                "SELECT * FROM templates WHERE id=$1 AND user_id=$2 AND is_active=TRUE",
                template_id,
                user_id
            )
            if not row:
                return None
            return Template(**dict(row))
    
    @staticmethod
    async def get_templates(user_id: str, skip: int = 0, limit: int = 100, search: Optional[str] = None) -> TemplateList:
        """
        Obtém a lista de templates de um usuário
        
        Args:
            user_id: ID do usuário proprietário dos templates
            skip: Número de registros para pular
            limit: Número máximo de registros para retornar
            search: Termo de busca opcional
            
        Returns:
            Lista de templates
        """
        async with await get_db_connection() as conn:
            where = ["user_id=$1", "is_active=TRUE"]
            params = [user_id]
            if search:
                where.append("(name ILIKE $2 OR content ILIKE $2)")
                params.append(f"%{search}%")
            where_sql = " AND ".join(where)
            query_sql = f"SELECT * FROM templates WHERE {where_sql} ORDER BY name OFFSET {skip} LIMIT {limit}"
            rows = await conn.fetch(query_sql, *params)
            count_sql = f"SELECT COUNT(*) FROM templates WHERE {where_sql}"
            total = await conn.fetchval(count_sql, *params)
            templates = [Template(**dict(r)) for r in rows]
            return TemplateList(templates=templates, total=total)
    
    @staticmethod
    async def update_template(user_id: str, template_id: str, template_in: TemplateUpdate) -> Template:
        """
        Atualiza um template
        
        Args:
            user_id: ID do usuário proprietário do template
            template_id: ID do template
            template_in: Dados a serem atualizados
            
        Returns:
            Template atualizado
            
        Raises:
            ValueError: Se o template não for encontrado
        """
        async with await get_db_connection() as conn:
            exists = await conn.fetchrow(
                "SELECT id FROM templates WHERE id=$1 AND user_id=$2 AND is_active=TRUE",
                template_id,
                user_id
            )
            if not exists:
                raise ValueError(f"Template não encontrado: {template_id}")
            data = template_in.dict(exclude_unset=True)
            data["updated_at"] = datetime.utcnow()
            set_parts = []
            values = []
            idx = 1
            for k, v in data.items():
                set_parts.append(f"{k}=${idx}")
                values.append(v)
                idx += 1
            values.extend([template_id, user_id])
            set_sql = ", ".join(set_parts)
            query = f"UPDATE templates SET {set_sql} WHERE id=${idx} AND user_id=${idx+1} RETURNING *"
            row = await conn.fetchrow(query, *values)
            return Template(**dict(row))
    
    @staticmethod
    async def delete_template(user_id: str, template_id: str) -> bool:
        """
        Remove um template (soft delete)
        
        Args:
            user_id: ID do usuário proprietário do template
            template_id: ID do template
            
        Returns:
            True se o template foi removido com sucesso
            
        Raises:
            ValueError: Se o template não for encontrado
        """
        async with await get_db_connection() as conn:
            row = await conn.fetchrow(
                "UPDATE templates SET is_active=FALSE, updated_at=$1 WHERE id=$2 AND user_id=$3 RETURNING id",
                datetime.utcnow(),
                template_id,
                user_id
            )
            if not row:
                raise ValueError(f"Template não encontrado: {template_id}")
            return True
